<a href="https://colab.research.google.com/github/dpsena210/machineLearn/blob/master/Projeto_final_Aprendizado_de_maquina.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Escopo do projeto

Para o nosso projeto será utilizado um data set sobre as características dos compradores de bicicleta encontrado no link https://www.kaggle.com/datasets/heeraldedhia/bike-buyers


1.    É problema de classificação binária.
2.    Deverá ser utilizado o método de **Validação Cruzada**
3.    Devemos trazer a **acurácia dos modelos**
4.    Devemos fazer a matriz de confusão
5.    A priori testaremos os seguintes modelos para descobrirmos o que terá melhor desempenho:



*   Árvore de decisão

*   Naive Bayes

*   Regressão Logística




Link introdutório sklearn: https://dadosaocubo.com/classificacao-com-scikit-learn/

O bloco abaixo importa as bibliotecas e pacotes que serão utilizadas para a implementação do modelo

In [165]:
import pandas as pd
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import ConfusionMatrixDisplay

from sklearn.metrics import classification_report
from sklearn.model_selection import train_test_split
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import confusion_matrix

from sklearn.metrics import accuracy_score
from sklearn.pipeline import Pipeline
#from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB

Nesta seção vamos importar o dataset utilizado para o notebook. Por limitação da ferramenta, o arquivo .csv utilizado deve ser inserido na seção de arquivos todas as vezes que o notebookfor inicializado. O arquivo para importação no notebook se encontra no caminho a seguir: https://drive.google.com/file/d/1tog969Iy5rny4cLPuse8kvQBYD6GD9M8/view?usp=share_link

In [ ]:
dados = pd.read_csv('https://raw.githubusercontent.com/dpsena210/machineLearn/master/bike_buyers_clean.csv')
dados

ID  Marital Status  Gender  Income  Children  Education  Home Owner  \
0    12496               1       1   40000         1          4           1   
1    24107               1       0   30000         3          3           1   
2    14177               1       0   80000         5          3           0   
3    24381               0       0   70000         0          4           1   
4    25597               0       0   30000         0          4           0   
..     ...             ...     ...     ...       ...        ...         ...   
995  23731               1       0   60000         2          2           1   
996  28672               0       0   70000         4          5           1   
997  11809               1       0   60000         2          4           1   
998  19664               0       0  100000         3          4           0   
999  12121               0       0   60000         3          2           1   

     Cars  Commute Distance  Age  Purchased Bike  
0       0                 1   42               0  
1       1                 1   43               0  
2       2                 5   60               0  
3       1                10   41               1  
4       0                 1   36               1  
..    ...               ...  ...             ...  
995     2                 5   54               1  
996     0                 5   35               1  
997     0                 1   38               1  
998     3                 2   38               0  
999     2                15   53               1  

[1000 rows x 11 columns]

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


Separando os dados em uma matriz de características, que são os dados utilizados para a predição e um vetor destino, que contém os valores target.

In [ ]:
variaveis = dados[['ID','Marital Status','Gender','Income','Children','Education','Home Owner','Cars','Commute Distance','Age']]
comprou_bike = dados['Purchased Bike']


In [ ]:
dados

ID  Marital Status  Gender  Income  Children  Education  Home Owner  \
0    12496               1       1   40000         1          4           1   
1    24107               1       0   30000         3          3           1   
2    14177               1       0   80000         5          3           0   
3    24381               0       0   70000         0          4           1   
4    25597               0       0   30000         0          4           0   
..     ...             ...     ...     ...       ...        ...         ...   
995  23731               1       0   60000         2          2           1   
996  28672               0       0   70000         4          5           1   
997  11809               1       0   60000         2          4           1   
998  19664               0       0  100000         3          4           0   
999  12121               0       0   60000         3          2           1   

     Cars  Commute Distance  Age  Purchased Bike  
0       0                 1   42               0  
1       1                 1   43               0  
2       2                 5   60               0  
3       1                10   41               1  
4       0                 1   36               1  
..    ...               ...  ...             ...  
995     2                 5   54               1  
996     0                 5   35               1  
997     0                 1   38               1  
998     3                 2   38               0  
999     2                15   53               1  

[1000 rows x 11 columns]

Aplicando o método train_test_split da biblioteca Scikit-Learn o conjunto de dados será separado em porções de testes(30%) e de treinamento(70%). O método realiza essa divisão de forma aleatória.\
**OBS:** O parametro *random-state* será configurado como 0 a fim de obter a mesma repartição de dados em todas as execuções.

# Escolha do modelo
O modelo a ser aplicado será decidido através da análise de desempenho. Como a proposta é utilizar um modelo linear para categorizar um perfil como comprador ou não de bicicletas, serão considerados 3 modelos distintos, sendo eles: Árvores de Classificação (CART), Naive Bayes (NB)

Para a análise serão utilizadas as métricas de acurácia e desvio padrão obtidos pelo metódo de validação cruzada k-fold, utilizando 10-folds

In [167]:
# definindo os folds
num_folds = 10
scoring = 'accuracy'

# Criando os modelos
modelo_cart = DecisionTreeClassifier()
modelo_nb = GaussianNB()
modelo_linear_regression = LogisticRegression()

# Avaliação dos modelos
kfold = KFold(n_splits=num_folds)
cv_results_cart = cross_val_score(DecisionTreeClassifier(), variaveis_train, comprou_bike_train, cv=kfold, scoring=scoring)
cv_results_nb = cross_val_score(GaussianNB(), variaveis_train, comprou_bike_train, cv=kfold, scoring=scoring)
cv_results_LR = cross_val_score(LogisticRegression(), variaveis_train, comprou_bike_train, cv=kfold, scoring=scoring)

print("Modelo Acurácia  DesvPad")

print("CART   %f  %f" % (cv_results_cart.mean(), cv_results_cart.std()))
print("NB     %f  %f" % (cv_results_nb.mean(), cv_results_nb.std()))
print("LR     %f  %f" % (cv_results_LR.mean(), cv_results_LR.std()))





Modelo Acurácia  DesvPad
CART   0.637143  0.060339
NB     0.618571  0.037823
LR     0.594286  0.042474


Com uma acurácia maior, escolhemos o modelo de Árvore de Decisão

# Escolha de hiperparametros


Escolha de hiperparametros: DecisionTreeClassifier(criterion, splitter, max_depth=None, min_samples_split=2, min_samples_leaf=1, min_weight_fraction_leaf=0.0, max_features=None, random_state=None, max_leaf_nodes=None, min_impurity_decrease=0.0, class_weight=None, ccp_alpha=0.0)

## Criterion

Como cada vez que realizavamos o modelo os valores diferiam e mostravam hora melhor em um, hora melhor outro parametro, fizemos uma iteração para encontrarmos o melhor hiperparametro para esse caso e após inúmeras repetições encontramos que o melhor o criterion foi o padrao , 'gini'. 

- MEDIA TOTAL HIPERPARAMETRO GINI :  0.637
- MEDIA TOTAL HIPERPARAMETRO ENTROPY :  0.631
- MEDIA TOTAL HIPERPARAMETRO LOG_LOSS :  0.629

Apesar de pouca diferença aos longos do testes o hiperparametro gini continuou apresentando os melhores resultados

In [133]:
class HyperparamCriterion:
  sum_hyperparam_gini = 0
  sum_hyperparam_entropy = 0
  sum_hyperparam_log_loss = 0

  def __init__(self, criterion_options):
    self.criterion_options = criterion_options

  def criterion_hyperparam_test(self, criterion, iterator1):
    modelo_cart_criterion = DecisionTreeClassifier(criterion=criterion)
    cv_results_cart_criterion = cross_val_score(modelo_cart_criterion, variaveis_train, comprou_bike_train, cv=kfold, scoring=scoring)
    mean_cv_results = cv_results_cart_criterion.mean()
    standard_deviation = cv_results_cart_criterion.std()

    
    if criterion == 'gini':
      self.sum_hyperparam_gini = self.sum_hyperparam_gini + mean_cv_results   
    if criterion == 'entropy':
      self.sum_hyperparam_entropy = self.sum_hyperparam_entropy + mean_cv_results 
    if criterion == 'log_loss':
      self.sum_hyperparam_log_loss = self.sum_hyperparam_log_loss + mean_cv_results 
   
    print("Iteração %i . hiperparametro criterion: %s  : Acurácia: %f , Desvio padrão: %f"%(iterator1, criterion, mean_cv_results, standard_deviation))

  def compare_hyperparams_criterion(self, iterations):
    for iterator1 in range(iterations):
      for criterion in criterion_options:
        acuracy = self.criterion_hyperparam_test(criterion, iterator1)
    print("MEDIA TOTAL HIPERPARAMETRO GINI : ",self.sum_hyperparam_gini/iterations)
    print("MEDIA TOTAL HIPERPARAMETRO ENTROPY : ",self.sum_hyperparam_entropy/iterations)    
    print("MEDIA TOTAL HIPERPARAMETRO LOG_LOSS : ",self.sum_hyperparam_log_loss/iterations)    

criterion_options = ['gini','entropy','log_loss']
testCriterion = HyperparamCriterion(criterion_options)
testCriterion.compare_hyperparams_criterion(100)

Iteração 0 . hiperparametro criterion: gini  : Acurácia: 0.632857 , Desvio padrão: 0.064222
Iteração 0 . hiperparametro criterion: entropy  : Acurácia: 0.635714 , Desvio padrão: 0.053548
Iteração 0 . hiperparametro criterion: log_loss  : Acurácia: 0.622857 , Desvio padrão: 0.053146
Iteração 1 . hiperparametro criterion: gini  : Acurácia: 0.651429 , Desvio padrão: 0.058971
Iteração 1 . hiperparametro criterion: entropy  : Acurácia: 0.635714 , Desvio padrão: 0.057588
Iteração 1 . hiperparametro criterion: log_loss  : Acurácia: 0.640000 , Desvio padrão: 0.047294
Iteração 2 . hiperparametro criterion: gini  : Acurácia: 0.631429 , Desvio padrão: 0.045535
Iteração 2 . hiperparametro criterion: entropy  : Acurácia: 0.627143 , Desvio padrão: 0.034670
Iteração 2 . hiperparametro criterion: log_loss  : Acurácia: 0.637143 , Desvio padrão: 0.047466
Iteração 3 . hiperparametro criterion: gini  : Acurácia: 0.637143 , Desvio padrão: 0.040000
Iteração 3 . hiperparametro criterion: entropy  : Acurácia:

## Splitter

Fizemos o mesmo para o Hiperparametro do splitter, devido ao mesmo fator, e o melhor resultado no caso foi o splitter='random':

 
- MEDIA TOTAL HIPERPARAMETRO BEST :  0.63
- MEDIA TOTAL HIPERPARAMETRO RANDOM :  0.64

Apesar de sutil, a diferença se repete ao longo de todos os testes

In [135]:
  

class HyperparamSplitter:
  sum_hyperparam_best = 0
  sum_hyperparam_random = 0

  def __init__(self, splitter_options):
    self.splitter_options = splitter_options

  def splitter_hyperparam_test(self,splitter,iterator2):

      modelo_cart_splitter = DecisionTreeClassifier(splitter=splitter)
      cv_results_cart_splitter = cross_val_score(modelo_cart_splitter, variaveis_train, comprou_bike_train, cv=kfold, scoring=scoring)
      
      mean_cv_results = cv_results_cart_splitter.mean();
      standard_deviation = cv_results_cart_splitter.std()

      if splitter == 'best':
        self.sum_hyperparam_best = self.sum_hyperparam_best + mean_cv_results
      if splitter == 'random':
        self.sum_hyperparam_random = self.sum_hyperparam_random +mean_cv_results 
      
      print("Iteração %i . hiperparametro splitter: %s  : Acurácia: %f , Desvio padrão: %f"%(iterator2, splitter, mean_cv_results, standard_deviation))

  def compare_splitters(self,iterations):
    for iterator2 in range(iterations):
        for splitter in self.splitter_options:
            self.splitter_hyperparam_test(splitter,iterator2)

    print("MEDIA TOTAL HIPERPARAMETRO BEST : ",self.sum_hyperparam_best/iterations)
    print("MEDIA TOTAL HIPERPARAMETRO RANDOM : ",self.sum_hyperparam_random/iterations)

splitter_options = ['best','random']

teste1 = HyperparamSplitter(splitter_options)

teste1.compare_splitters(100)


Iteração 0 . hiperparametro splitter: best  : Acurácia: 0.617143 , Desvio padrão: 0.063824
Iteração 0 . hiperparametro splitter: random  : Acurácia: 0.644286 , Desvio padrão: 0.049672
Iteração 1 . hiperparametro splitter: best  : Acurácia: 0.638571 , Desvio padrão: 0.066409
Iteração 1 . hiperparametro splitter: random  : Acurácia: 0.655714 , Desvio padrão: 0.053624
Iteração 2 . hiperparametro splitter: best  : Acurácia: 0.637143 , Desvio padrão: 0.043425
Iteração 2 . hiperparametro splitter: random  : Acurácia: 0.641429 , Desvio padrão: 0.038571
Iteração 3 . hiperparametro splitter: best  : Acurácia: 0.645714 , Desvio padrão: 0.051824
Iteração 3 . hiperparametro splitter: random  : Acurácia: 0.641429 , Desvio padrão: 0.051289
Iteração 4 . hiperparametro splitter: best  : Acurácia: 0.635714 , Desvio padrão: 0.067688
Iteração 4 . hiperparametro splitter: random  : Acurácia: 0.638571 , Desvio padrão: 0.056079
Iteração 5 . hiperparametro splitter: best  : Acurácia: 0.637143 , Desvio padrão

## Max Depth

Para o hiperparametro maxdepth, fizemos uma serie de repeticoes que serao as linhas da matriz e cada numero de decisoes da arvore a ser testado serão as colunas, dessa forma podemos tirar a media e ver qual delas se saem melhor quando repetida 10 vezes, os testes variaram de 1 a 64. Obtivemos que o hiperparametro max_depth quando variado entre os inteiros tem alta latencia, enquanto o valor padrão None mantém uma regularidade maior
- max_depth:None.  Acurácia :  0.64
- Demais no array impresso 

In [146]:
class HyperparamMaxDepth:
  #matriz para acumulo de testes, as linhas representam as repeticoes de cada hyperparametro max_depth e as colunas o numero do max_depth propriamente dito
  
  
  def __init__(self,number_of_decisions, times_to_repeat):
      self.times_to_repeat = times_to_repeat
      self.number_of_decisions = number_of_decisions
      self.matrix = [[None for j in range(number_of_decisions)] for i in range(times_to_repeat)]
      
      #array com todas as medias
      self.arrayMeansMaxDepth = [0 for h in range(number_of_decisions)]


  def max_depth_hyperparam_test(self, decisions, row):
    modelo_cart_max_depth = DecisionTreeClassifier(max_depth=decisions+1)
    cv_results_cart_max_depth = cross_val_score(modelo_cart_max_depth, variaveis_train, comprou_bike_train, cv=kfold, scoring=scoring)
    
    mean_cv_results = cv_results_cart_max_depth.mean()
    standard_deviation_cv = cv_results_cart_max_depth.std()

    self.matrix[row][decisions] = mean_cv_results

  def max_depth_default(self):
    modelo_cart_max_depth = DecisionTreeClassifier(max_depth=None)
    cv_results_cart_max_depth = cross_val_score(modelo_cart_max_depth, variaveis_train, comprou_bike_train, cv=kfold, scoring=scoring)
    
    mean_cv_results = cv_results_cart_max_depth.mean()
    standard_deviation_cv = cv_results_cart_max_depth.std()

    print("max_depth:None,  Acurácia : ", mean_cv_results)

  #Criação da matriz 
  def generateDecisionsMatrix(self):
    for i in range(self.times_to_repeat):
      for decisions in range(self.number_of_decisions):
        self.max_depth_hyperparam_test(decisions,i)
      print(i)  
      decisions =0  

  def allResultsMeans(self):
    for i in range(len(self.matrix)):
      for j in range(len(self.matrix[i])):
        self.arrayMeansMaxDepth[j] = self.arrayMeansMaxDepth[j] + self.matrix[i][j]


    for k in range(len(self.arrayMeansMaxDepth)):
      self.arrayMeansMaxDepth[k] = self.arrayMeansMaxDepth[k]/self.times_to_repeat  
    print("Média das iterações para os max_dephs de 1 a 64",self.arrayMeansMaxDepth)
    return self.arrayMeansMaxDepth

#Algoritmo com 10 repetiçoes max def indo de 1 a 64
testeMaxDepth = HyperparamMaxDepth(64,10)
testeMaxDepth.generateDecisionsMatrix()
arrayMeansMaxDepth = testeMaxDepth.allResultsMeans()

testeMaxDepth.max_depth_default()

0
1
2
3
4
5
6
7
8
9
Média das iterações para os max_dephs de 1 a 64 [0.62, 0.6328571428571427, 0.6042857142857142, 0.5878571428571429, 0.6, 0.6054285714285714, 0.6291428571428572, 0.6332857142857145, 0.6227142857142858, 0.6268571428571429, 0.6327142857142858, 0.6335714285714287, 0.6312857142857143, 0.6328571428571428, 0.6314285714285716, 0.6345714285714286, 0.6321428571428571, 0.638, 0.639, 0.6362857142857143, 0.6410000000000001, 0.6362857142857143, 0.6394285714285715, 0.6371428571428572, 0.6317142857142859, 0.639, 0.6367142857142858, 0.6404285714285713, 0.6365714285714287, 0.6351428571428572, 0.6357142857142858, 0.6358571428571429, 0.636, 0.6338571428571429, 0.6374285714285716, 0.6354285714285713, 0.6362857142857143, 0.6365714285714287, 0.6357142857142858, 0.6330000000000001, 0.6335714285714286, 0.6338571428571429, 0.6322857142857143, 0.6395714285714286, 0.6395714285714286, 0.6394285714285715, 0.6361428571428571, 0.6347142857142857, 0.6387142857142857, 0.634, 0.6372857142857145, 0.635

Verificando os valores de maior eficácia, percebemos que sempre muda quando o teste é rodado novamente.

In [147]:
melhor_acuracia = 0

#extraindo a maior acurácia e sua posição +1 para encontrarmos o número do max_depth

for i in range(len(arrayMeansMaxDepth)):
  if arrayMeansMaxDepth[i] > melhor_acuracia:
    melhor_acuracia = arrayMeansMaxDepth[i]
    maior_depth = i

print("Temos que a maior acuracia foi: ", melhor_acuracia, " para o número de depths = ", maior_depth+1)    


Temos que a maior acuracia foi:  0.6410000000000001  para o número de depths =  21


## Min samples splits

Para o hiperparametro min_samples_splits fizemos o teste com os valores :
MEDIA TOTAL HIPERPARAMETRO MIN_SAMPLES_SPLITS = 0.1 :  0.627
MEDIA TOTAL HIPERPARAMETRO MIN_SAMPLES_SPLITS = 1 :  0.555
MEDIA TOTAL HIPERPARAMETRO MIN_SAMPLES_SPLITS = 2 :  0.644
MEDIA TOTAL HIPERPARAMETRO MIN_SAMPLES_SPLITS = 10 :  0.635

  E obtivemos o melhor desempenho para o valor 2 com uma acuracia de 64,39 % 

In [149]:
class HyperparamMinSamplesSplit:
  sum_hyperparam_0_1 = 0
  sum_hyperparam_1 = 0
  sum_hyperparam_2 = 0
  sum_hyperparam_10 = 0
  def __init__(self, min_samples_split_options):
    self.min_samples_split_options = min_samples_split_options

  def min_samples_split_hyperparam_test(self,min_samples_split,iterator2):

      modelo_cart_min_samples_split = DecisionTreeClassifier(min_samples_split=min_samples_split,max_depth=None,splitter="random")
      cv_results_cart_min_samples_split = cross_val_score(modelo_cart_min_samples_split, variaveis_train, comprou_bike_train, cv=kfold, scoring=scoring)
      
      mean_cv_results = cv_results_cart_min_samples_split.mean();
      standard_deviation = cv_results_cart_min_samples_split.std()

      if min_samples_split == 0.1:
        self.sum_hyperparam_0_1 = self.sum_hyperparam_0_1 + mean_cv_results
      if min_samples_split == 1.0:
        self.sum_hyperparam_1 = self.sum_hyperparam_1 +mean_cv_results 
      if min_samples_split == 2:
        self.sum_hyperparam_2 = self.sum_hyperparam_2 +mean_cv_results 
      if min_samples_split == 10:
        self.sum_hyperparam_10 = self.sum_hyperparam_10 +mean_cv_results 

      
      print("Iteração %i . hiperparametro splitter: %s  : Acurácia: %f , Desvio padrão: %f"%(iterator2, min_samples_split, mean_cv_results, standard_deviation))

  def compare_min_samples_split(self,times_to_repeat):
    for iterator3 in range(times_to_repeat):
        for min_sample_split in self.min_samples_split_options:
            self.min_samples_split_hyperparam_test(min_sample_split,iterator3)

    print("MEDIA TOTAL HIPERPARAMETRO MIN_SAMPLES_SPLITS = 0.1 : ",self.sum_hyperparam_0_1/times_to_repeat)
    print("MEDIA TOTAL HIPERPARAMETRO MIN_SAMPLES_SPLITS = 1 : ",self.sum_hyperparam_1/times_to_repeat)
    print("MEDIA TOTAL HIPERPARAMETRO MIN_SAMPLES_SPLITS = 2 : ",self.sum_hyperparam_2/times_to_repeat)

    print("MEDIA TOTAL HIPERPARAMETRO MIN_SAMPLES_SPLITS = 10 : ",self.sum_hyperparam_10/times_to_repeat)

min_sample_split_options =[0.1,1.0,10,2]
test_min_samples_split = HyperparamMinSamplesSplit(min_sample_split_options)    
test_min_samples_split.compare_min_samples_split(100)

Iteração 0 . hiperparametro splitter: 0.1  : Acurácia: 0.622857 , Desvio padrão: 0.040507
Iteração 0 . hiperparametro splitter: 1.0  : Acurácia: 0.552857 , Desvio padrão: 0.057870
Iteração 0 . hiperparametro splitter: 10  : Acurácia: 0.651429 , Desvio padrão: 0.067673
Iteração 0 . hiperparametro splitter: 2  : Acurácia: 0.640000 , Desvio padrão: 0.044630
Iteração 1 . hiperparametro splitter: 0.1  : Acurácia: 0.624286 , Desvio padrão: 0.063904
Iteração 1 . hiperparametro splitter: 1.0  : Acurácia: 0.475714 , Desvio padrão: 0.035600
Iteração 1 . hiperparametro splitter: 10  : Acurácia: 0.678571 , Desvio padrão: 0.061029
Iteração 1 . hiperparametro splitter: 2  : Acurácia: 0.652857 , Desvio padrão: 0.041429
Iteração 2 . hiperparametro splitter: 0.1  : Acurácia: 0.630000 , Desvio padrão: 0.037498
Iteração 2 . hiperparametro splitter: 1.0  : Acurácia: 0.574286 , Desvio padrão: 0.086142
Iteração 2 . hiperparametro splitter: 10  : Acurácia: 0.637143 , Desvio padrão: 0.052372
Iteração 2 . hipe

## Min samples leaf

Com o hiperparametro min_samples_leaf padrão =  1 também obtivemos a melhor acuracia:

- MEDIA TOTAL HIPERPARAMETRO MIN_SAMPLES_LEAF = 0.1 :  0.609
- MEDIA TOTAL HIPERPARAMETRO MIN_SAMPLES_LEAF = 0.5 :  0.523
- MEDIA TOTAL HIPERPARAMETRO MIN_SAMPLES_LEAF = 1 :  0.645
- MEDIA TOTAL HIPERPARAMETRO MIN_SAMPLES_LEAF = 5 :  0.632

In [151]:
class HyperparamMinSamplesLeaf:
  sum_hyperparam_0_1 = 0
  sum_hyperparam_0_5 = 0
  sum_hyperparam_1 = 0
  sum_hyperparam_5 = 0
  def __init__(self, min_samples_leaf_options):
    self.min_samples_leaf_options = min_samples_leaf_options

  def min_samples_leaf_hyperparam_test(self,min_samples_leaf,iterator2):

      modelo_cart_min_samples_leaf = DecisionTreeClassifier(min_samples_split=2,min_samples_leaf=min_samples_leaf,max_depth=28,splitter="random")
      cv_results_cart_min_samples_leaf = cross_val_score(modelo_cart_min_samples_leaf, variaveis_train, comprou_bike_train, cv=kfold, scoring=scoring)
      
      mean_cv_results = cv_results_cart_min_samples_leaf.mean();
      standard_deviation = cv_results_cart_min_samples_leaf.std()

      if min_samples_leaf == 0.1:
        self.sum_hyperparam_0_1 = self.sum_hyperparam_0_1 + mean_cv_results
      if min_samples_leaf == 0.5:
        self.sum_hyperparam_0_5 = self.sum_hyperparam_0_5 +mean_cv_results 
      if min_samples_leaf == 1:
        self.sum_hyperparam_1 = self.sum_hyperparam_1 +mean_cv_results 
      if min_samples_leaf == 5:
        self.sum_hyperparam_5 = self.sum_hyperparam_5 +mean_cv_results 

      
      print("Iteração %i . hiperparametro splitter: %s  : Acurácia: %f , Desvio padrão: %f"%(iterator2, min_samples_leaf, mean_cv_results, standard_deviation))

  def compare_min_samples_leaf(self,times_to_repeat):
    for iterator4 in range(times_to_repeat):
        for min_samples_leaf in self.min_samples_leaf_options:
            self.min_samples_leaf_hyperparam_test(min_samples_leaf,iterator4)

    print("MEDIA TOTAL HIPERPARAMETRO MIN_SAMPLES_LEAF = 0.1 : ",self.sum_hyperparam_0_1/times_to_repeat)
    print("MEDIA TOTAL HIPERPARAMETRO MIN_SAMPLES_LEAF = 0.5 : ",self.sum_hyperparam_0_5/times_to_repeat)
    print("MEDIA TOTAL HIPERPARAMETRO MIN_SAMPLES_LEAF = 1 : ",self.sum_hyperparam_1/times_to_repeat)

    print("MEDIA TOTAL HIPERPARAMETRO MIN_SAMPLES_LEAF = 5 : ",self.sum_hyperparam_5/times_to_repeat)

min_sample_leaf_options =[0.1,0.5,1,5]
test_min_samples_leaf = HyperparamMinSamplesLeaf(min_sample_leaf_options)    
test_min_samples_leaf.compare_min_samples_leaf(100)

Iteração 0 . hiperparametro splitter: 0.1  : Acurácia: 0.588571 , Desvio padrão: 0.059864
Iteração 0 . hiperparametro splitter: 0.5  : Acurácia: 0.527143 , Desvio padrão: 0.055861
Iteração 0 . hiperparametro splitter: 1  : Acurácia: 0.621429 , Desvio padrão: 0.035714
Iteração 0 . hiperparametro splitter: 5  : Acurácia: 0.617143 , Desvio padrão: 0.047724
Iteração 1 . hiperparametro splitter: 0.1  : Acurácia: 0.595714 , Desvio padrão: 0.065481
Iteração 1 . hiperparametro splitter: 0.5  : Acurácia: 0.521429 , Desvio padrão: 0.059676
Iteração 1 . hiperparametro splitter: 1  : Acurácia: 0.638571 , Desvio padrão: 0.045647
Iteração 1 . hiperparametro splitter: 5  : Acurácia: 0.635714 , Desvio padrão: 0.056152
Iteração 2 . hiperparametro splitter: 0.1  : Acurácia: 0.600000 , Desvio padrão: 0.033806
Iteração 2 . hiperparametro splitter: 0.5  : Acurácia: 0.521429 , Desvio padrão: 0.059676
Iteração 2 . hiperparametro splitter: 1  : Acurácia: 0.641429 , Desvio padrão: 0.043542
Iteração 2 . hiperpa

## Max features

Para o Parametro max_features os resultados ficaram incoclusivos e variando, por isso optamos por deixar o valor padrao = None 

In [152]:
numero_de_colunas = 9


In [159]:

class HyperparamMaxFeatures:  
  
  def __init__(self,number_of_features, times_to_repeat):
      self.times_to_repeat = times_to_repeat
      self.number_of_features = number_of_features
      self.matrix2 = [[None for j in range(number_of_features)] for i in range(times_to_repeat)]
      
      #array com todas as medias
      self.arrayMeans2 = [0 for h in range(number_of_features)]


  def max_features_hyperparam_test(self, features, row):
    modelo_cart_max_features = DecisionTreeClassifier(min_samples_split=2,min_samples_leaf=1,max_depth=28,splitter="random", max_features=features+1)
    cv_results_cart_max_features = cross_val_score(modelo_cart_max_features, variaveis_train, comprou_bike_train, cv=kfold, scoring=scoring)
    
    mean_cv_results = cv_results_cart_max_features.mean()
    standard_deviation_cv = cv_results_cart_max_features.std()

    self.matrix2[row][features] = mean_cv_results

  #Criação da matriz 
  def generateDecisionsMatrix(self):
    for i in range(self.times_to_repeat):
      for features in range(self.number_of_features):
        self.max_features_hyperparam_test(features,i)
      print(i)  
      decisions =0  

  def allResultsMeans(self):
    for i in range(len(self.matrix2)):
      for j in range(len(self.matrix2[i])):
        self.arrayMeans2[j] = self.arrayMeans2[j] + self.matrix2[i][j]


    for k in range(len(self.arrayMeans2)):
      self.arrayMeans2[k] = self.arrayMeans2[k]/self.times_to_repeat
    print(self.arrayMeans2)
    return self.arrayMeans2
testeMaxFeatures = HyperparamMaxFeatures(numero_de_colunas,100)
testeMaxFeatures.generateDecisionsMatrix()
arrayMaxFeatures = testeMaxFeatures.allResultsMeans()

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
[0.6197285714285715, 0.6278000000000001, 0.631357142857143, 0.6382571428571431, 0.6411714285714286, 0.6396142857142856, 0.6420142857142856, 0.643257142857143, 0.6444285714285715]


In [160]:


melhor_acuracia_feature = 0
for q in range(len(arrayMaxFeatures)):
  if arrayMaxFeatures[q] > melhor_acuracia_feature:
    melhor_acuracia_feature = arrayMaxFeatures[q]
    maior_feature = q+1
print(arrayMaxFeatures)
print("melhor acuracia acontece com o a quantidade de features: ",maior_feature, " , cujo valor é : ",melhor_acuracia_feature)

[0.6197285714285715, 0.6278000000000001, 0.631357142857143, 0.6382571428571431, 0.6411714285714286, 0.6396142857142856, 0.6420142857142856, 0.643257142857143, 0.6444285714285715]
melhor acuracia acontece com o a quantidade de features:  9  , cujo valor é :  0.6444285714285715


## Min weight fraction leaf

Para o hiperparametro min_weight_fraction_leaf  escolhemos um intervalo de 0.0 até 0.5 e percebemos que além de novamente o valor padrão '0.0' levar a melhor:

- MEDIA TOTAL HIPERPARAMETRO MIN_WEIGHT_FRACTION_LEAF = 0.0 :  0.643
- MEDIA TOTAL HIPERPARAMETRO MIN_WEIGHT_FRACTION_LEAF = 0.1 :  0.606
- MEDIA TOTAL HIPERPARAMETRO MIN_WEIGHT_FRACTION_LEAF = 0.2 :  0.589
- MEDIA TOTAL HIPERPARAMETRO MIN_WEIGHT_FRACTION_LEAF = 0.3 :  0.576
- MEDIA TOTAL HIPERPARAMETRO MIN_WEIGHT_FRACTION_LEAF = 0.4 :  0.574
- MEDIA TOTAL HIPERPARAMETRO MIN_WEIGHT_FRACTION_LEAF = 0.5 :  0.523

In [161]:
class HyperparamMinWeightFractionLeaf:
  sum_hyperparam_0_1 = 0
  sum_hyperparam_0_2 = 0
  sum_hyperparam_0_3 = 0
  sum_hyperparam_0_4 = 0
  sum_hyperparam_0_5 = 0
  sum_hyperparam_0_0 = 0

  def __init__(self, min_weight_fraction_leaf_options):
    self.min_weight_fraction_leaf_options = min_weight_fraction_leaf_options

  def min_weight_fraction_leaf_hyperparam_test(self,min_weight_fraction_leaf,iterator5):

      modelo_cart_min_weight_fraction_leaf = DecisionTreeClassifier(min_weight_fraction_leaf=min_weight_fraction_leaf, min_samples_split=2, min_samples_leaf=1, max_depth=28, splitter="random")
      cv_results_cart_min_weight_fraction_leaf = cross_val_score(modelo_cart_min_weight_fraction_leaf, variaveis_train, comprou_bike_train, cv=kfold, scoring=scoring)
      
      mean_cv_results = cv_results_cart_min_weight_fraction_leaf.mean();
      standard_deviation = cv_results_cart_min_weight_fraction_leaf.std()

      if min_weight_fraction_leaf == 0.0:
        self.sum_hyperparam_0_0 = self.sum_hyperparam_0_0 +mean_cv_results
      if min_weight_fraction_leaf == 0.1:
        self.sum_hyperparam_0_1 = self.sum_hyperparam_0_1 + mean_cv_results
      if min_weight_fraction_leaf == 0.2:
        self.sum_hyperparam_0_2 = self.sum_hyperparam_0_2 +mean_cv_results 
      if min_weight_fraction_leaf == 0.3:
        self.sum_hyperparam_0_3 = self.sum_hyperparam_0_3 +mean_cv_results 
      if min_weight_fraction_leaf == 0.4:
        self.sum_hyperparam_0_4 = self.sum_hyperparam_0_4 +mean_cv_results 
      if min_weight_fraction_leaf == 0.5:
        self.sum_hyperparam_0_5 = self.sum_hyperparam_0_5 +mean_cv_results 
      

      
      print("Iteração %i . hiperparametro splitter: %s  : Acurácia: %f , Desvio padrão: %f"%(iterator5, min_weight_fraction_leaf, mean_cv_results, standard_deviation))

  def compare_max_leaf_nodes(self,times_to_repeat):
    for iterator5 in range(times_to_repeat):
        for min_weight_fraction_leaf in self.min_weight_fraction_leaf_options:
            self.min_weight_fraction_leaf_hyperparam_test(min_weight_fraction_leaf,iterator5)

    print("MEDIA TOTAL HIPERPARAMETRO MIN_WEIGHT_FRACTION_LEAF = 0.0 : ",self.sum_hyperparam_0_0/times_to_repeat)
    print("MEDIA TOTAL HIPERPARAMETRO MIN_WEIGHT_FRACTION_LEAF = 0.1 : ",self.sum_hyperparam_0_1/times_to_repeat)
    print("MEDIA TOTAL HIPERPARAMETRO MIN_WEIGHT_FRACTION_LEAF = 0.2 : ",self.sum_hyperparam_0_2/times_to_repeat)
    print("MEDIA TOTAL HIPERPARAMETRO MIN_WEIGHT_FRACTION_LEAF = 0.3 : ",self.sum_hyperparam_0_3/times_to_repeat)
    print("MEDIA TOTAL HIPERPARAMETRO MIN_WEIGHT_FRACTION_LEAF = 0.4 : ",self.sum_hyperparam_0_4/times_to_repeat)
    print("MEDIA TOTAL HIPERPARAMETRO MIN_WEIGHT_FRACTION_LEAF = 0.5 : ",self.sum_hyperparam_0_5/times_to_repeat)


min_weight_fraction_leaf_options =[0.0, 0.1,0.2,0.3,0.4,0.5]
test_min_weight_fraction_leaf = HyperparamMinWeightFractionLeaf(min_weight_fraction_leaf_options)    
test_min_weight_fraction_leaf.compare_max_leaf_nodes(100)

Iteração 0 . hiperparametro splitter: 0.0  : Acurácia: 0.657143 , Desvio padrão: 0.035571
Iteração 0 . hiperparametro splitter: 0.1  : Acurácia: 0.581429 , Desvio padrão: 0.046969
Iteração 0 . hiperparametro splitter: 0.2  : Acurácia: 0.577143 , Desvio padrão: 0.069752
Iteração 0 . hiperparametro splitter: 0.3  : Acurácia: 0.567143 , Desvio padrão: 0.043354
Iteração 0 . hiperparametro splitter: 0.4  : Acurácia: 0.552857 , Desvio padrão: 0.035022
Iteração 0 . hiperparametro splitter: 0.5  : Acurácia: 0.521429 , Desvio padrão: 0.059676
Iteração 1 . hiperparametro splitter: 0.0  : Acurácia: 0.640000 , Desvio padrão: 0.037688
Iteração 1 . hiperparametro splitter: 0.1  : Acurácia: 0.602857 , Desvio padrão: 0.071085
Iteração 1 . hiperparametro splitter: 0.2  : Acurácia: 0.584286 , Desvio padrão: 0.063391
Iteração 1 . hiperparametro splitter: 0.3  : Acurácia: 0.558571 , Desvio padrão: 0.037498
Iteração 1 . hiperparametro splitter: 0.4  : Acurácia: 0.574286 , Desvio padrão: 0.039279
Iteração 1

## Max leaf nodes

O Hiperparametro max_leaf_nodes Leva valores inteiros, testaremos de 2 a 50. Nesse caso foi inconclusivo entao deixarmos o padrão


In [163]:
class HyperparamMaxLeafNodes:
  #matriz para acumulo de testes, as linhas representam as repeticoes de cada hyperparametro max_depth e as colunas o numero do max_depth propriamente dito
  
  
  def __init__(self,number_of_max_leaf_nodes, times_to_repeat):
      self.times_to_repeat = times_to_repeat
      self.number_of_max_leaf_nodes = number_of_max_leaf_nodes
      self.matrix3 = [[None for j in range(number_of_max_leaf_nodes)] for i in range(times_to_repeat)]
      
      #array com todas as medias
      self.arrayMeans3 = [0 for h in range(number_of_max_leaf_nodes)]


  def max_leaf_nodes_hyperparam_test(self, max_leaf_nodes, row):
    modelo_cart_max_leaf_nodes = DecisionTreeClassifier(min_samples_split=2,min_samples_leaf=1,max_depth=28,splitter="random", max_leaf_nodes=max_leaf_nodes+2)
    cv_results_cart_max_leaf_nodes = cross_val_score(modelo_cart_max_leaf_nodes, variaveis_train, comprou_bike_train, cv=kfold, scoring=scoring)
    
    mean_cv_results = cv_results_cart_max_leaf_nodes.mean()
    standard_deviation_cv = cv_results_cart_max_leaf_nodes.std()

    self.matrix3[row][max_leaf_nodes] = mean_cv_results

  #Criação da matriz 
  def generateDecisionsMatrix(self):
    for j in range(self.times_to_repeat):
      for max_leaf_nodes in range(self.number_of_max_leaf_nodes):
        self.max_leaf_nodes_hyperparam_test(max_leaf_nodes,j)
      print(j)  
      decisions =0  

  def allResultsMeans(self):
    for i in range(len(self.matrix3)):
      for j in range(len(self.matrix3[i])):
        self.arrayMeans3[j] = self.arrayMeans3[j] + self.matrix3[i][j]


    for k in range(len(self.arrayMeans3)):
      self.arrayMeans3[k] = self.arrayMeans3[k]/self.times_to_repeat
    print(self.arrayMeans3)
    return self.arrayMeans3
testeMaxLeafNodes = HyperparamMaxLeafNodes(51,20)
testeMaxLeafNodes.generateDecisionsMatrix()
arrayMaxLeafNodes = testeMaxLeafNodes.allResultsMeans()

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
[0.5563571428571429, 0.5945714285714285, 0.6032142857142857, 0.6057142857142856, 0.6152142857142857, 0.6053571428571429, 0.6151428571428571, 0.6162857142857142, 0.6115, 0.616142857142857, 0.6199999999999999, 0.6244999999999999, 0.6230714285714286, 0.6130000000000001, 0.6229285714285713, 0.6180714285714286, 0.622142857142857, 0.6232857142857142, 0.6306428571428571, 0.6250714285714285, 0.6270714285714286, 0.6306428571428572, 0.6278571428571429, 0.6250714285714285, 0.6299999999999999, 0.6285, 0.6272857142857143, 0.6323571428571428, 0.6242857142857143, 0.6334285714285712, 0.6277857142857143, 0.6264285714285714, 0.6307142857142858, 0.6302857142857142, 0.6336428571428571, 0.6320714285714285, 0.6315, 0.6362857142857143, 0.6320000000000001, 0.6365000000000001, 0.6314285714285715, 0.6329999999999999, 0.639357142857143, 0.6327142857142855, 0.6315714285714286, 0.633, 0.6412142857142858, 0.6403571428571428, 0.6367142857142857, 0.6413571428571428, 0

In [95]:
melhor_acuracia_max_leaf_nodes = 0
for w in range(len(arrayMaxLeafNodes)):
  if arrayMaxLeafNodes[w] > melhor_acuracia_max_leaf_nodes:
    melhor_acuracia_max_leaf_nodes = arrayMaxLeafNodes[w]
    melhor_max_leaf_nodes_number = w+2
print(arrayMaxFeatures)
print("melhor acuracia acontece com o hiperparametro max_leaf_nodes = ",melhor_max_leaf_nodes_number, " , com acuracia de : ",melhor_acuracia_max_leaf_nodes)

[0.6211428571428572, 0.634857142857143, 0.6357142857142859, 0.6382857142857142, 0.6351428571428571, 0.636, 0.6422857142857142, 0.6497142857142857, 0.6454285714285714, 0.666, 0.6491428571428571, 0.6594285714285715, 0.6534285714285715, 0.6485714285714285, 0.6328571428571429, 0.6328571428571429, 0.6394285714285715, 0.6568571428571429, 0.635142857142857, 0.6562857142857143, 0.6594285714285715, 0.6448571428571428, 0.6502857142857144, 0.6408571428571429, 0.6417142857142858, 0.6517142857142858, 0.648, 0.6442857142857144, 0.642, 0.646, 0.6465714285714286, 0.652857142857143, 0.6522857142857144, 0.6425714285714286, 0.6448571428571429, 0.6505714285714286, 0.6562857142857144, 0.6431428571428571, 0.6439999999999999, 0.6462857142857142, 0.6460000000000001, 0.6371428571428571, 0.6357142857142858, 0.6514285714285715, 0.6468571428571428, 0.6294285714285713, 0.6497142857142857, 0.646857142857143, 0.6457142857142857, 0.6442857142857144]
melhor acuracia acontece com o hiperparametro max_leaf_nodes =  51  

## Random state

Com o hiperparametro random_state, seguimos o mesmo procedimento para encontrar qual a melhor opção e encontramos o valor 37 e uma acurácia de aproximadamente 69,3 %

In [63]:
class HyperparamRandomState:
  #matriz para acumulo de testes, as linhas representam as repeticoes de cada hyperparametro max_depth e as colunas o numero do max_depth propriamente dito
  
  
  def __init__(self,number_of_random_states, times_to_repeat):
      self.times_to_repeat = times_to_repeat
      self.number_of_random_states = number_of_random_states
      self.matrix4 = [[None for j in range(number_of_random_states)] for i in range(times_to_repeat)]
      
      #array com todas as medias
      self.arrayMeans4 = [0 for h in range(number_of_random_states)]


  def random_states_hyperparam_test(self, random_states, row):
    modelo_cart_random_states = DecisionTreeClassifier(min_samples_split=2,min_samples_leaf=1,max_depth=28,splitter="random", random_state=random_states)
    cv_results_cart_random_states = cross_val_score(modelo_cart_random_states, variaveis_train, comprou_bike_train, cv=kfold, scoring=scoring)
    
    mean_cv_results = cv_results_cart_random_states.mean()
    standard_deviation_cv = cv_results_cart_random_states.std()

    self.matrix4[row][random_states] = mean_cv_results

  #Criação da matriz 
  def generateDecisionsMatrix(self):
    for j in range(self.times_to_repeat):
      for random_states in range(self.number_of_random_states):
        self.random_states_hyperparam_test(random_states,j)
      print(j)  
      decisions =0  

  def allResultsMeans(self):
    for i in range(len(self.matrix4)):
      for j in range(len(self.matrix4[i])):
        self.arrayMeans4[j] = self.arrayMeans4[j] + self.matrix4[i][j]


    for k in range(len(self.arrayMeans4)):
      self.arrayMeans4[k] = self.arrayMeans4[k]/self.times_to_repeat
    print(self.arrayMeans4)
    return self.arrayMeans4
test_random_states = HyperparamRandomState(51,20)
test_random_states.generateDecisionsMatrix()
arrayRandomStates = test_random_states.allResultsMeans()

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
[0.6485714285714282, 0.6657142857142856, 0.6799999999999998, 0.6442857142857145, 0.6414285714285713, 0.67, 0.6328571428571428, 0.6214285714285717, 0.657142857142857, 0.6185714285714284, 0.647142857142857, 0.6742857142857143, 0.6542857142857146, 0.6314285714285716, 0.6414285714285715, 0.6371428571428571, 0.6257142857142858, 0.6285714285714283, 0.6271428571428572, 0.6271428571428572, 0.6199999999999998, 0.6642857142857145, 0.6285714285714286, 0.6142857142857143, 0.6571428571428571, 0.6385714285714285, 0.66, 0.6257142857142858, 0.657142857142857, 0.6714285714285713, 0.657142857142857, 0.615714285714286, 0.6685714285714286, 0.67, 0.6514285714285714, 0.692857142857143, 0.6400000000000001, 0.6642857142857145, 0.6271428571428572, 0.6485714285714289, 0.6571428571428571, 0.657142857142857, 0.6585714285714287, 0.6414285714285715, 0.6271428571428572, 0.6285714285714283, 0.6528571428571427, 0.6628571428571431, 0.6300000000000002, 0.6542857142857144

In [71]:
melhor_acuracia_random_state = 0
for w in range(len(arrayRandomStates)):
  if arrayRandomStates[w] > melhor_acuracia_random_state:
    melhor_acuracia_random_state = arrayRandomStates[w]
    melhor_random_state = w
print(arrayMaxFeatures)
print("melhor acuracia acontece com o hiperparametro random_state = ",melhor_random_state, " , com acuracia de : ",melhor_acuracia_random_state)

[0.6211428571428572, 0.634857142857143, 0.6357142857142859, 0.6382857142857142, 0.6351428571428571, 0.636, 0.6422857142857142, 0.6497142857142857, 0.6454285714285714, 0.666, 0.6491428571428571, 0.6594285714285715, 0.6534285714285715, 0.6485714285714285, 0.6328571428571429, 0.6328571428571429, 0.6394285714285715, 0.6568571428571429, 0.635142857142857, 0.6562857142857143, 0.6594285714285715, 0.6448571428571428, 0.6502857142857144, 0.6408571428571429, 0.6417142857142858, 0.6517142857142858, 0.648, 0.6442857142857144, 0.642, 0.646, 0.6465714285714286, 0.652857142857143, 0.6522857142857144, 0.6425714285714286, 0.6448571428571429, 0.6505714285714286, 0.6562857142857144, 0.6431428571428571, 0.6439999999999999, 0.6462857142857142, 0.6460000000000001, 0.6371428571428571, 0.6357142857142858, 0.6514285714285715, 0.6468571428571428, 0.6294285714285713, 0.6497142857142857, 0.646857142857143, 0.6457142857142857, 0.6442857142857144]
melhor acuracia acontece com o hiperparametro random_state =  35  , 

## Min impurity decrease

Para o min_impurity_decrease, temos duas opções o valor calculado ou o padrão, como veremos em seguida o valor padrão apresenta maior acurácia

In [74]:


clf = DecisionTreeClassifier(min_impurity_decrease=0.01)
clf.fit(variaveis_test, comprou_bike_test)

# impureza da raiz da árvore de decisão
root_impurity = clf.tree_.impurity[0]

# impureza dos nós filhos esquerdo e direito da raiz da árvore de decisão
left_impurity = clf.tree_.impurity[clf.tree_.children_left[0]]
right_impurity = clf.tree_.impurity[clf.tree_.children_right[0]]

# número de amostras nos nós filhos esquerdo e direito da raiz da árvore de decisão
n_left = clf.tree_.n_node_samples[clf.tree_.children_left[0]]
n_right = clf.tree_.n_node_samples[clf.tree_.children_right[0]]

# Fórmula do min_impurity_decrease encontrada no sklearn doc
weighted_average_impurity = (n_left * left_impurity + n_right * right_impurity) / (n_left + n_right)
weighted_samples = n_left + n_right

min_impurity_decrease = (root_impurity - weighted_average_impurity) / weighted_samples

print("Valor de min_impurity_decrease calculado é:", min_impurity_decrease)



Valor de min_impurity_decrease calculado é: 6.407897603485849e-05


Calculando as acurácias:

In [76]:
#Acurácia para o impurity decrease calculado
model_with_min_impurity_calculated = DecisionTreeClassifier(min_samples_split=2,min_samples_leaf=1,max_depth=28,splitter="random", random_state=35, min_impurity_decrease=0.00006407897603485849)
cv_results_cart_random_states = cross_val_score(model_with_min_impurity_calculated, variaveis_train, comprou_bike_train, cv=kfold, scoring=scoring)

mean_cv_results = cv_results_cart_random_states.mean()
standard_deviation_cv = cv_results_cart_random_states.std()

mean_cv_results

0.6857142857142857

In [77]:
#Acurácia para o impurity decrease padrão
model_with_min_impurity_calculated = DecisionTreeClassifier(min_samples_split=2,min_samples_leaf=1,max_depth=28,splitter="random", random_state=35, min_impurity_decrease=0.0)
cv_results_cart_random_states = cross_val_score(model_with_min_impurity_calculated, variaveis_train, comprou_bike_train, cv=kfold, scoring=scoring)

mean_cv_results = cv_results_cart_random_states.mean()
standard_deviation_cv = cv_results_cart_random_states.std()

mean_cv_results

0.6928571428571428

## class_weight

In [103]:
dados

ID  Marital Status  Gender  Income  Children  Education  Home Owner  \
0    12496               1       1   40000         1          4           1   
1    24107               1       0   30000         3          3           1   
2    14177               1       0   80000         5          3           0   
3    24381               0       0   70000         0          4           1   
4    25597               0       0   30000         0          4           0   
..     ...             ...     ...     ...       ...        ...         ...   
995  23731               1       0   60000         2          2           1   
996  28672               0       0   70000         4          5           1   
997  11809               1       0   60000         2          4           1   
998  19664               0       0  100000         3          4           0   
999  12121               0       0   60000         3          2           1   

     Cars  Commute Distance  Age  Purchased Bike  
0       0                 1   42               0  
1       1                 1   43               0  
2       2                 5   60               0  
3       1                10   41               1  
4       0                 1   36               1  
..    ...               ...  ...             ...  
995     2                 5   54               1  
996     0                 5   35               1  
997     0                 1   38               1  
998     3                 2   38               0  
999     2                15   53               1  

[1000 rows x 11 columns]

In [109]:

 
def class_weight_hyperparam_test(class_weight):

  modelo_cart_class_weight = DecisionTreeClassifier(class_weight=class_weight, min_samples_split=2, min_samples_leaf=1, max_depth=28, splitter="random",random_state=35, min_impurity_decrease=0.0)
  cv_results_cart_class_weight = cross_val_score(modelo_cart_class_weight, variaveis_train, comprou_bike_train, cv=kfold, scoring=scoring)

  mean_cv_results = cv_results_cart_class_weight.mean();
  standard_deviation = cv_results_cart_class_weight.std()

  print(" hiperparametro class weight: %s  : Acurácia: %f , Desvio padrão: %f"%( class_weight, mean_cv_results, standard_deviation))



Como temos as colunas:
-	Marital Status (Estado civil)
- Gender (Gênero)
-	Income (Renda)
- Children (Filhos)
-	Education (Formação)
-	Home Owner (Proprietários de casa)
-	Cars (Carros)
-	Commute Distance (Distancia de deslocamento)
-	Age	 (Idade)

Testaremos as combinações de peso conforme nossa percepção em 5 variações, também testamos com o valor "balanced" e o valor padrão(None):


In [113]:
combinacao1 = {0:1,1:2,2:2,3:1,4:1,5:1,6:4,7:4,8:3 }
combinacao2 = {0:2,1:3,2:3,3:3,4:2,5:5,6:6,7:5,8:4 }
combinacao3 = {0:1,1:4,2:6,3:1,4:1,5:2,6:10,7:10,8:3 }
combinacao4 = {0:5,1:4,2:1,3:8,4:8,5:7,6:3,7:4,8:5 }
combinacao5 = {0:10,1:7,2:5,3:4,4:4,5:3,6:7,7:7,8:9 }
balanced ="balanced"
padrao = None

lista_combinacoes = [combinacao1,combinacao2,combinacao3,combinacao4,combinacao5,balanced, padrao]


for combinacao in lista_combinacoes:
  class_weight_hyperparam_test(combinacao)



 hiperparametro class weight: {0: 1, 1: 2, 2: 2, 3: 1, 4: 1, 5: 1, 6: 4, 7: 4, 8: 3}  : Acurácia: 0.660000 , Desvio padrão: 0.062204
 hiperparametro class weight: {0: 2, 1: 3, 2: 3, 3: 3, 4: 2, 5: 5, 6: 6, 7: 5, 8: 4}  : Acurácia: 0.658571 , Desvio padrão: 0.058011
 hiperparametro class weight: {0: 1, 1: 4, 2: 6, 3: 1, 4: 1, 5: 2, 6: 10, 7: 10, 8: 3}  : Acurácia: 0.651429 , Desvio padrão: 0.037904
 hiperparametro class weight: {0: 5, 1: 4, 2: 1, 3: 8, 4: 8, 5: 7, 6: 3, 7: 4, 8: 5}  : Acurácia: 0.647143 , Desvio padrão: 0.053088
 hiperparametro class weight: {0: 10, 1: 7, 2: 5, 3: 4, 4: 4, 5: 3, 6: 7, 7: 7, 8: 9}  : Acurácia: 0.658571 , Desvio padrão: 0.057658
 hiperparametro class weight: balanced  : Acurácia: 0.658571 , Desvio padrão: 0.049672
 hiperparametro class weight: None  : Acurácia: 0.692857 , Desvio padrão: 0.055051


Observamos que o valor None nos trás uma melhor acurácia

## Ccp alpha

Encontramos que para o ccp_alpha o valor d 0.0  ate 0.001 se saem melhor
- MEDIA TOTAL HIPERPARAMETRO CCP_ALPHA = 0.0 :  0.693
- MEDIA TOTAL HIPERPARAMETRO CCP_ALPHA = 0.0001 :  0.693
- MEDIA TOTAL HIPERPARAMETRO CCP_ALPHA = 0.001 :  0.693
- MEDIA TOTAL HIPERPARAMETRO CCP_ALPHA = 0.01 :  0.601
- MEDIA TOTAL HIPERPARAMETRO CCP_ALPHA = 0.1 :  0.521
- MEDIA TOTAL HIPERPARAMETRO CCP_ALPHA = 1 :  0.521
- MEDIA TOTAL HIPERPARAMETRO CCP_ALPHA = 10 :  0.521
- MEDIA TOTAL HIPERPARAMETRO CCP_ALPHA = 100 :  0.521
- MEDIA TOTAL HIPERPARAMETRO CCP_ALPHA = 1000 :  0.521

In [131]:
class HyperparamCcpAlpha:
  sum_hyperparam_0 = 0
  sum_hyperparam_1 = 0
  sum_hyperparam_2 = 0
  sum_hyperparam_3 = 0
  sum_hyperparam_4 = 0
  sum_hyperparam_5 = 0
  sum_hyperparam_6 = 0
  sum_hyperparam_7 = 0
  sum_hyperparam_8 = 0
  def __init__(self, ccp_alpha_options):
    self.ccp_alpha_options = ccp_alpha_options

  def ccp_alpha_options_hyperparam_test(self,ccp_alpha):

      modelo_cart_ccp_alpha = DecisionTreeClassifier(min_samples_split=2,max_depth=28,splitter="random",random_state=35,ccp_alpha=ccp_alpha)
      cv_results_cart_ccp_alpha = cross_val_score(modelo_cart_ccp_alpha, variaveis_train, comprou_bike_train, cv=kfold, scoring=scoring)
      
      mean_cv_results = cv_results_cart_ccp_alpha.mean();
      standard_deviation = cv_results_cart_ccp_alpha.std()

      if ccp_alpha == 0.0:
        self.sum_hyperparam_0 =  mean_cv_results
      if ccp_alpha == 0.0001:
        self.sum_hyperparam_1 =  mean_cv_results
      if ccp_alpha == 0.001:
        self.sum_hyperparam_2 = mean_cv_results 
      if ccp_alpha == 0.01:
        self.sum_hyperparam_3 = mean_cv_results 
      if ccp_alpha ==  0.1:
        self.sum_hyperparam_4 = mean_cv_results 
      if ccp_alpha ==  1:
        self.sum_hyperparam_5 = mean_cv_results 
      if ccp_alpha ==  10:
        self.sum_hyperparam_6 = mean_cv_results
      if ccp_alpha ==  100:
        self.sum_hyperparam_7 = mean_cv_results 
      if ccp_alpha ==  1000:
        self.sum_hyperparam_8 = mean_cv_results   

      
      print("hiperparametro ccp_alpha: %s  : Acurácia: %f , Desvio padrão: %f"%(ccp_alpha, mean_cv_results, standard_deviation))

  def compare_ccp_alpha(self):
    
    for ccp_alpha in self.ccp_alpha_options:
      self.ccp_alpha_options_hyperparam_test(ccp_alpha)
    print("MEDIA TOTAL HIPERPARAMETRO CCP_ALPHA = 0.0 : ",self.sum_hyperparam_0)

    print("MEDIA TOTAL HIPERPARAMETRO CCP_ALPHA = 0.0001 : ",self.sum_hyperparam_1)
    print("MEDIA TOTAL HIPERPARAMETRO CCP_ALPHA = 0.001 : ",self.sum_hyperparam_2)
    print("MEDIA TOTAL HIPERPARAMETRO CCP_ALPHA = 0.01 : ",self.sum_hyperparam_3)

    print("MEDIA TOTAL HIPERPARAMETRO CCP_ALPHA = 0.1 : ",self.sum_hyperparam_4)
    print("MEDIA TOTAL HIPERPARAMETRO CCP_ALPHA = 1 : ",self.sum_hyperparam_5)
    print("MEDIA TOTAL HIPERPARAMETRO CCP_ALPHA = 10 : ",self.sum_hyperparam_6)
    print("MEDIA TOTAL HIPERPARAMETRO CCP_ALPHA = 100 : ",self.sum_hyperparam_7)
    print("MEDIA TOTAL HIPERPARAMETRO CCP_ALPHA = 1000 : ",self.sum_hyperparam_8)



ccp_alpha_options =[0.0,0.0001,0.001,0.01,0.1,1,10,100,1000]
test_ccp_alpha = HyperparamCcpAlpha(ccp_alpha_options)    
test_ccp_alpha.compare_ccp_alpha()

hiperparametro ccp_alpha: 0.0  : Acurácia: 0.692857 , Desvio padrão: 0.055051
hiperparametro ccp_alpha: 0.0001  : Acurácia: 0.692857 , Desvio padrão: 0.055051
hiperparametro ccp_alpha: 0.001  : Acurácia: 0.692857 , Desvio padrão: 0.050810
hiperparametro ccp_alpha: 0.01  : Acurácia: 0.601429 , Desvio padrão: 0.056946
hiperparametro ccp_alpha: 0.1  : Acurácia: 0.521429 , Desvio padrão: 0.059676
hiperparametro ccp_alpha: 1  : Acurácia: 0.521429 , Desvio padrão: 0.059676
hiperparametro ccp_alpha: 10  : Acurácia: 0.521429 , Desvio padrão: 0.059676
hiperparametro ccp_alpha: 100  : Acurácia: 0.521429 , Desvio padrão: 0.059676
hiperparametro ccp_alpha: 1000  : Acurácia: 0.521429 , Desvio padrão: 0.059676
MEDIA TOTAL HIPERPARAMETRO CCP_ALPHA = 0.0 :  0.6928571428571428
MEDIA TOTAL HIPERPARAMETRO CCP_ALPHA = 0.0001 :  0.6928571428571428
MEDIA TOTAL HIPERPARAMETRO CCP_ALPHA = 0.001 :  0.6928571428571428
MEDIA TOTAL HIPERPARAMETRO CCP_ALPHA = 0.01 :  0.6014285714285714
MEDIA TOTAL HIPERPARAMETRO C

**Colocar uns gráficos para a validação**

# Treinando o modelo
Com a escolha do modelo de *TreeDecisionClassifier* vamos realizar o treinamento utilizando o modelo já inicializado na seção anterior

In [177]:
modelClassifier = DecisionTreeClassifier(class_weight=None, min_samples_split=2, min_samples_leaf=1, max_depth=28, splitter="random",random_state=35, min_impurity_decrease=0.0, ccp_alpha=0.0)

modelClassifier.fit(variaveis_train, comprou_bike_train)



DecisionTreeClassifier(max_depth=28, random_state=35, splitter='random')

# Predições com os dados de testes

In [180]:
y_pred = modelClassifier.predict(variaveis_test)
print(y_pred)





[1 0 1 0 0 0 0 1 0 1 1 0 1 1 1 0 1 1 1 0 0 1 1 0 1 0 0 1 0 0 1 0 0 1 0 0 0
 0 1 1 1 1 1 0 0 0 0 1 0 1 0 0 1 1 1 0 0 1 1 0 1 1 0 1 1 1 1 1 0 0 0 0 1 1
 0 1 1 0 0 0 1 1 1 1 1 0 0 1 0 1 0 1 1 1 1 0 1 1 1 0 1 0 1 1 1 0 1 1 0 0 1
 0 1 1 1 1 1 0 0 1 0 1 0 0 1 0 0 0 0 0 0 0 0 0 1 1 0 0 0 0 1 1 0 1 0 0 1 1
 1 0 1 1 0 1 1 0 0 0 0 1 0 1 0 0 1 1 1 0 1 0 1 1 0 1 1 1 0 1 1 1 1 1 0 0 0
 1 1 1 0 1 0 0 1 0 0 0 1 1 0 1 0 0 1 0 1 1 0 0 1 0 1 0 0 1 1 1 1 0 1 1 0 0
 0 0 1 0 0 1 1 0 1 1 0 0 0 1 0 0 0 0 0 0 0 1 0 1 0 0 1 0 1 0 0 1 0 1 1 0 0
 0 1 0 1 1 0 1 0 0 0 0 0 1 1 1 0 1 1 0 1 0 1 0 0 1 0 1 1 1 0 0 0 1 1 0 0 0
 1 0 0 0]


#Acurácia do modelo

In [182]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

print("Acurácia:", accuracy_score(comprou_bike_test, y_pred))
print("Matriz de confusão:\n", confusion_matrix(comprou_bike_test, y_pred))
print("Relatório de classificação:\n", classification_report(comprou_bike_test, y_pred))

Acurácia: 0.6833333333333333
Matriz de confusão:
 [[106  48]
 [ 47  99]]
Relatório de classificação:
               precision    recall  f1-score   support

           0       0.69      0.69      0.69       154
           1       0.67      0.68      0.68       146

    accuracy                           0.68       300
   macro avg       0.68      0.68      0.68       300
weighted avg       0.68      0.68      0.68       300

